pkg/reconciler/mtnamespace/namespace.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mtnamespace
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/eventing/pkg/reconciler/mtnamespace/resources"
+
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	eventinglisters "knative.dev/eventing/pkg/client/listers/eventing/v1beta1"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"knative.dev/eventing/pkg/apis/eventing/v1alpha1"
+	"knative.dev/eventing/pkg/apis/eventing/v1beta1"
+	"knative.dev/eventing/pkg/logging"
+	"knative.dev/eventing/pkg/reconciler"
+	"knative.dev/pkg/controller"
+)
+
+const (
+	namespaceReconciled       = "NamespaceReconciled"
+	namespaceReconcileFailure = "NamespaceReconcileFailure"
+
+	// Name of the corev1.Events emitted from the reconciliation process.
+	brokerCreated = "BrokerCreated"
+)
+
+var (
+	brokerGVK = v1alpha1.SchemeGroupVersion.WithKind("Broker")
+)
+
+type Reconciler struct {
+	*reconciler.Base
+
+	// listers index properties about resources
+	namespaceLister corev1listers.NamespaceLister
+	brokerLister    eventinglisters.BrokerLister
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Namespace resource
+// with the current status of the resource.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logging.FromContext(ctx).Error("invalid resource key")
+		return nil
+	}
+
+	// Get the namespace resource with this namespace/name
+	original, err := r.namespaceLister.Get(name)
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Error("namespace key in work queue no longer exists")
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	if original.Labels[resources.InjectionLabelKey] == resources.InjectionDisabledLabelValue {
```

Comment:
So, this means that to opt-out, you have to create the namespace with this label in it which means you can't just do:
kubectl create ns foobar
but have to create the object (or is there a way to do this?)

I'm fine with getting this in as is for now, but perhaps we should use the configmap to control this behaviour for the opt-out as a follow on?
Also, could you add an issue for adding the docs for this feature?

Thoughts?

Reply:
Yeah, I don't see an option to label namespaces in `kubectl create ns foo`.  I think if we were going to go the ConfigMap route we'd want consistency with ST Broker.  Env vars seem to be popular for configuring this.

How about making the default behavior configurable via `DEFAULT_BROKER_INJECTION` and having the two reconcilers default it differently?

re: docs

I'm happy to follow up with a PR after this lands.  I was thinking we'd fold the "enable" tab into the ST Broker instructions, and combine this with the MT Broker instructions.  I'll start on that and we can discuss there?

Reply:
https://github.com/knative/docs/pull/2325